include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h

```diff
@@ -570,12 +570,24 @@ struct __radix_sort_onesweep_kernel
                                                __local_tid * __bin_width * sizeof(_LocOffsetT),
                                            __scan<_LocOffsetT, _LocOffsetT>(__thread_grf_hist_summary));
 
-            // 1.3 Copy the histogram at the region designited for synchronization between work-groups.
+            // 1.3. Copy the histogram at the region designited for synchronization between work-groups.
             // Set the status as "updated".
             if (__wg_id != 0)
-                __dpl_esimd::__block_store<::std::uint32_t, __bin_width>(
+            {
+                // Write the histogram explicitly to L2, bypassing L1.
+                // L2 is assumed to be coherent, thus other work-groups can read from there if possible
+                __dpl_esimd::__ens::lsc_block_store<::std::uint32_t, __bin_width,
+                                                    __dpl_esimd::__ens::lsc_data_size::default_size,
+                                                    __dpl_esimd::__ens::cache_hint::write_through,
+                                                    __dpl_esimd::__ens::cache_hint::write_back>(
                     __p_this_group_hist + __local_tid * __bin_width, __thread_grf_hist_summary | __hist_updated);
+            }
         }
+        // Make sure the histogram updated at the step 1.3 is visible to other groups
+        // The histogram data above is explicitly written to L2, which is assumed to be coherent: no need to flush it
+        __dpl_esimd::__ns::fence<__dpl_esimd::__ns::memory_kind::global,
+                                 __dpl_esimd::__ns::fence_flush_op::none,
+                                 __dpl_esimd::__ns::fence_scope::gpu>();
         __dpl_esimd::__ns::barrier();
```

Comment:
> The documentation for barrier() says "Memory ordering is also guaranteed by this instruction."

That's a good question. The documentation says nothing about the type of the memory. It is reasonable to assume any memory type, but this is what I see in the implementation:
```c++
 __ESIMD_API void barrier() {
   __esimd_fence(fence_mask::global_coherent_fence | fence_mask::local_barrier);
   __esimd_barrier();
 }
```
The description of the flags:

|  |                                                   |
|-----------------------|--------------------------------------------------------------|
| global_coherent_fence | "Commit enable" - wait for fence to complete before continuing. |
| local_barrier         | Issue SLM memory barrier only. If not set, the memory barrier is global. |

The behaviour is not clear. I assume that the barrier works only for local memory operations. I will reach the ESIMD developers to clarify it.

Let me answer other complex question:

> Is the new fence necessary because of the inter-work-group loads that follow?

Yes

>  Is the danger stemming from the possibility that those loads could potentially read an element from global memory whose actual value is stuck in some other work-group's L1 cache that has not been flushed?

Exactly. I assume that work-groups can delay writing to global memory or the L2 cache until any point in time, unless a fence is used.

> And that barrier() might not perform a write back?

Yes, this is my concern. I will try to confirm it, though.

Reply:
I've read the documentation of `esimd::barrier` more carefully. 
> Performs barrier synchronization for all threads within the same thread group... Memory ordering is also guaranteed by this instruction.

It is fair to expect global memory operations order within a work-group. It is still not sufficient for cross-work-group synchronization through global memory. 

Upd.:
I've reached ESIMD developers, and it seems the barrier enforces local memory order only.

Reply:
Thanks for the thorough explanation.

> Upd.: I've reached ESIMD developers, and it seems the barrier enforces local memory order only.

I agree that the proposed changes are necessary with this confirmation that barriers are for local memory only.